### PR TITLE
K8SPSMDB-845: delete backup finalizer fails with PBM v2 

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPSMDB-845](https://jira.percona.com/browse/K8SPSMDB-845) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPSMDB-845

**DESCRIPTION**
---
**Problem:**
*When we run backup on different storages and then delete the first one, `delete-backup` finalizer fails*

**Cause:**
*Every time we create a backup we set a new pbm config, which contains info about storage.  `delete-backup` finalizer uses pbm to delete backups. When we delete the backup with other storage, we don't set a new pbm config with info about the needed storage. This way pbm, uses storage of the last created backup and tries to delete backup on the wrong storage*

**Solution:**
*We should set a new pbm config every time the `delete-backup` finalizer used*

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?
- [x] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [x] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [ ] Are E2E tests passing?
- [x] Are unit tests passing?
- [x] Are linting tests passing?
- [ ] Is an E2E test/test case added for the new feature/change?
- [ ] Are unit tests added where appropriate?
- [ ] Are OpenShift compare files changed for E2E tests (`compare/*-oc.yml`)?

**Config/Logging/Testability**
- [ ] Are all needed new/changed options added to default YAML files?
- [ ] Are the manifests (crd/bundle) regenerated if needed?
- [ ] Did we add proper logging messages for operator actions?
- [ ] Did we ensure compatibility with the previous version or cluster upgrade process?
- [ ] Does the change support oldest and newest supported MongoDB version?
- [ ] Does the change support oldest and newest supported Kubernetes version?

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| cross-site-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/cross-site-sharded.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/init-deploy.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/one-pod.log) |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/storage.log) |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/upgrade-consistency.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/demand-backup.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/self-healing-chaos.log) |
| ignore-labels-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/ignore-labels-annotations.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/limits.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/scaling.log) |
| monitoring-2-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/monitoring-2-0.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/scheduled-backup.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/operator-self-healing-chaos.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/security-context.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/arbiter.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/rs-shard-migration.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/demand-backup-sharded.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/service-per-pod.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/liveness.log) |
| upgrade | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/upgrade.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/smart-update.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/version-service.log) |
| upgrade-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/upgrade-sharded.log) |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/pitr.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/users.log) |
| pitr-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/pitr-sharded.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/data-sharded.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/non-voting.log) |
| demand-backup-eks-credentials | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/demand-backup-eks-credentials.log) |
| data-at-rest-encryption | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/8b67291d/data-at-rest-encryption.log) |
| We run 30 out of 30|

commit: https://github.com/percona/percona-server-mongodb-operator/pull/1137/commits/8b67291db52e2b688dca079927cd2599e891519b
image: `perconalab/percona-server-mongodb-operator:PR-1137-8b67291d`

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/init-deploy.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/one-pod.log) |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/upgrade-consistency.log) |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/storage.log) |
| cross-site-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/cross-site-sharded.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/demand-backup.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/self-healing-chaos.log) |
| ignore-labels-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/ignore-labels-annotations.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/limits.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/scaling.log) |
| monitoring-2-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/monitoring-2-0.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/operator-self-healing-chaos.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/security-context.log) |
| scheduled-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/scheduled-backup.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/rs-shard-migration.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/arbiter.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/service-per-pod.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/liveness.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/smart-update.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/demand-backup-sharded.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/version-service.log) |
| upgrade | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/upgrade.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/users.log) |
| upgrade-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/upgrade-sharded.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/data-sharded.log) |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/pitr.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/non-voting.log) |
| demand-backup-eks-credentials | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/demand-backup-eks-credentials.log) |
| data-at-rest-encryption | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/data-at-rest-encryption.log) |
| pitr-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/pitr-sharded.log) |
| We run 30 out of 30|

commit: https://github.com/percona/percona-server-mongodb-operator/pull/1137/commits/126bc15628301d732cce207bd9c86f15f240ba64
image: `perconalab/percona-server-mongodb-operator:PR-1137-126bc156`

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/one-pod.log) |
| ignore-labels-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/ignore-labels-annotations.log) |
| monitoring-2-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/monitoring-2-0.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/arbiter.log) |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/storage.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/init-deploy.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/service-per-pod.log) |
| cross-site-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/cross-site-sharded.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/self-healing-chaos.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/limits.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/liveness.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/operator-self-healing-chaos.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/scaling.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/smart-update.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/security-context.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/version-service.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/rs-shard-migration.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/users.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/data-sharded.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/non-voting.log) |
| demand-backup-eks-credentials | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/demand-backup-eks-credentials.log) |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/upgrade-consistency.log) |
| data-at-rest-encryption | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/data-at-rest-encryption.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/demand-backup.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/scheduled-backup.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/demand-backup-sharded.log) |
| upgrade | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/upgrade.log) |
| upgrade-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/upgrade-sharded.log) |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/pitr.log) |
| pitr-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-1137/126bc156/pitr-sharded.log) |
| We run 30 out of 30|

commit: https://github.com/percona/percona-server-mongodb-operator/pull/1137/commits/126bc15628301d732cce207bd9c86f15f240ba64
image: `perconalab/percona-server-mongodb-operator:PR-1137-126bc156`

**Review by hors — APPROVED** [review]

File: pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go [hunk 1]
```diff
@@ -341,16 +341,39 @@ func (r *ReconcilePerconaServerMongoDBBackup) checkFinalizers(ctx context.Contex
...
+				case cr.Status.Azure != nil:
+					storage.Type = psmdbv1.BackupStorageAzure
+					storage.Azure = *cr.Status.Azure
+				}
+
+				priorities, err := b.pbm.GetPriorities(ctx, r.client, cluster)
+				if err != nil {
+					return errors.Wrap(err, "get priorities")
+				}
+				err = b.pbm.SetConfig(ctx, storage, b.spec.PITR, priorities)
```
Comment (tplavcic) [inline pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go:368]: So if I understand correctly we set storage to the one which we need to delete, do we set it back to the one which was before?

**Review by tplavcic — CHANGES_REQUESTED** [review]

File: e2e-tests/demand-backup/run [hunk 2]
```diff
@@ -118,7 +118,42 @@ fi
...
-	#	check_backup_deletion "https://storage.googleapis.com/operator-testing/${backup_dest_gcp}" "gcp-cs"
+	check_backup_deletion "https://storage.googleapis.com/operator-testing/${backup_dest_gcp}" "gcp-cs"
+	check_backup_deletion "https://engk8soperators.blob.core.windows.net/operator-testing/${backup_dest_azure}" "azure-blob"
+fi
+
+desc 'checking backup deletion without cr'
+run_backup minio
+if [ -z "$SKIP_BACKUPS_TO_AWS_GCP_AZURE" ]; then
+	run_backup aws-s3
+	run_backup gcp-cs
```
Comment (nmarukovich) [inline e2e-tests/demand-backup/run:129]: Why we can do it step by step:
```
run_backup minio
wait_backup "$backup_name_minio"
if [ -z "$SKIP_BACKUPS_TO_AWS_GCP_AZURE" ]; then
	run_backup aws-s3
	run_backup gcp-cs
	run_backup azure-blob
        wait_backup "$backup_name_aws"
	wait_backup "$backup_name_gcp"
	wait_backup "$backup_name_azure"
fi

```

File: pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go [hunk 3]
```diff
@@ -307,7 +307,7 @@ func getPBMBackupMeta(cr *psmdbv1.PerconaServerMongoDBBackup) *pbm.BackupMeta {
 	return meta
 }
 
-func (r *ReconcilePerconaServerMongoDBBackup) checkFinalizers(ctx context.Context, cr *psmdbv1.PerconaServerMongoDBBackup, b *Backup) error {
+func (r *ReconcilePerconaServerMongoDBBackup) checkFinalizers(ctx context.Context, cr *psmdbv1.PerconaServerMongoDBBackup, cluster *psmdbv1.PerconaServerMongoDB, b *Backup) error {
```
Comment (inelpandzic) [inline pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go:310]: I would maybe call this as `db`, it is more descriptive I think, and `cluster` is pretty overloaded.

File: e2e-tests/demand-backup/run (hunk 2 above)
Comment (inelpandzic) [reply e2e-tests/demand-backup/run:129]: We can, but this way then we run_backup minio and wait_backup minio immediately and then run backups on other storages and wait for them.
If we want first to run backups on all storages and then wait for all of them, then we can't do it like this.

**Review by inelpandzic — APPROVED** [review]
Besides Tomislavs comment, this looks good to me.

File: pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go (hunk 1 above)
Comment (pooknull) [reply pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go:368]: I don't think that we should set it to the previous one. When we do a restore, we set a proper config here: https://github.com/percona/percona-server-mongodb-operator/blob/1235fb3b92af16e9fc2691e5a2a5a2ceb882e228/pkg/controller/perconaservermongodbrestore/perconaservermongodbrestore_controller.go#L228

Comment (hors) [reply pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go:368]: > I don't think that we should set it to the previous one. When we do a restore, we set a proper config here:
> 
> https://github.com/percona/percona-server-mongodb-operator/blob/1235fb3b92af16e9fc2691e5a2a5a2ceb882e228/pkg/controller/perconaservermongodbrestore/perconaservermongodbrestore_controller.go#L228

But what about the backup?

Comment (egegunes) [reply pkg/controller/perconaservermongodbbackup/perconaservermongodbbackup_controller.go:368]: we set config before backup too: https://github.com/percona/percona-server-mongodb-operator/blob/main/pkg/controller/perconaservermongodbbackup/backup.go#L48

**Review by egegunes — APPROVED** [review]
I'm OK with these changes to fix the finalizer issue in v1.14 but I strongly believe that we should stop depending on PBM and delete the backup files by using the appropriate client of object storage. This code depends replset pods to be up and running and in case of namespace deletion (as we do in our e2e tests) `PerconaServerMongoDB` will be deleted without waiting backup finalizers at all and some backup objects will block deletion because of that.

We also need to block operator termination (like we do in PXC operator) in case there are finalizers in `PerconaServerMongoDB` and `PerconaServerMongoDBBackup` objects to give operator a chance to run finalizers properly.

**Pull request merged**